### ISSUE TITLE
Allow definition of custom functions / classes at document level

### ISSUE THREAD

**Issue by jperon** [gristlabs/grist-core#386]
This is an idea I had several times, but which is very well illustrated by [this case](https://community.getgrist.com/t/how-to-create-a-formula-to-show-accumulated-amount-per-account/1861/6): it’s sometimes necessary to use custom functions / classes, and while they can be defined within column formulas, this leads to code duplication. It would be very handy if we had a way to add python code that would go to the top of the code, just after imports:

```python
import grist
from functions import *       # global uppercase functions
import datetime, math, re     # modules commonly needed in formulas

<<<HERE WOULD GO CUSTOM CODE>>>
```

In the case linked, this would allow to define `SList` at the document level, not within the function body.

**Comment by jperon** [comment]
While thinking about it, it would also be practical to have the option to define class functions for each table.

**Comment by dmw24** [comment]
This would be super helpful keeping my DB functions clean and easy to update

**Comment by emanuelegissi** [comment]
Yes, indeed very useful

**Comment by paulfitz** [comment]
It does seems a good idea. For UI, I could imagine an editable region in the Code Viewer. The data engine would need a bit of tweaking to make it use the code, and update correctly when it changes. If anyone wants to take a shot at it, it does seem an achievable feature, though not trivial.

**Comment by kamilmodest** [comment]
While it's not a solution, unfortunately, but a workaround that could be helpful for someone who found this page via googling:

https://community.getgrist.com/t/install-third-party-libraries/1087/3

TL;DR; you need to create your own Dockerfile with injecting all the necessary additional Python libraries in it.

**Comment by tayflo** [comment]
An easy and handy workaround:

- Create a dedicated table, for instance named "Function"

- In each column, define your function (or class, or constant, or whatever python object you want) and return it

```py
def say_hello(name: str) -> str:
  return f'Hello {name}!'

return say_hello
```

- In another table, retrieve the function from the table and call it as you want it (for instance here, pass the column "Name" as a parameter)

```py
Function.lookupOne().say_hello($Name)
```

You can declare stuff as complex as you want this way, and reuse it wherever you like within your document.

Note: Renaming some variables/properties or such within the declaration might break some downstream usages, as Grist handles Table/Column renaming, but not Python code refactoring. For instance renaming a method in a class declaration will break the method calls in another table: as you export only the class, Grist does not know that its inner logic has changed.

See also: [tomnitschke/gristruc](https://github.com/tomnitschke/gristruc)

---

Demo document available [here](https://docs.getgrist.com/wkQj8rkrCBip/demofunctions-definition?utm_id=share-doc), full document code right below

```py
import grist
from functions import *       # global uppercase functions
import datetime, math, re     # modules commonly needed in formulas


@grist.UserTable
class Function:

  def say_hello(rec, table):
    def say_hello(name: str) -> str:
      return f'Hello {name}!'

    return say_hello

  def Person(rec, table):
    class Person:
      def __init__(self, name: str, age: int):
        self.name = name
        self.age = age
    
      def is_above_18(self) -> bool:
        return self.age >= 18

    return Person


@grist.UserTable
class Person:
  Name = grist.Text()
  Age = grist.Numeric()

  def Hello(rec, table):
    return Function.lookupOne().say_hello(rec.Name)

  def Person_object(rec, table):
    return Function.lookupOne().Person(rec.Name, rec.Age)

  def is_above_18(rec, table):
    return rec.Person_object.is_above_18()
```
